src/mscorlib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs

```diff
@@ -319,20 +319,50 @@ public static AsyncTaskMethodBuilder<TResult> Create()
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.stateMachine);
             }
 
-            // Run the MoveNext method within a copy-on-write ExecutionContext scope.
-            // This allows us to undo any ExecutionContext changes made in MoveNext,
+            // enregistrer variables with 0 post-fix so they can be used in registers without EH forcing them to stack
+            // Capture references to Thread Contexts
+            Thread currentThread0 = Thread.CurrentThread;
+            Thread currentThread = currentThread0;
+            ref ExecutionContext currentExecutionCtxRef0 = ref currentThread0.ExecutionContext;
+            ExecutionContext previousExecutionCtx0 = currentExecutionCtxRef0;
+
+            // Store current ExecutionContext and SynchronizationContext as "previousXxx".
+            // This allows us to restore them and undo any Context changes made in stateMachine.MoveNext
             // so that they won't "leak" out of the first await.
+            ExecutionContext previousExecutionCtx = previousExecutionCtx0;
+            SynchronizationContext previousSyncCtx = currentThread0.SynchronizationContext;
 
-            Thread currentThread = Thread.CurrentThread;
-            ExecutionContextSwitcher ecs = default(ExecutionContextSwitcher);
             try
             {
-                ExecutionContext.EstablishCopyOnWriteScope(currentThread, ref ecs);
                 stateMachine.MoveNext();
             }
             finally
             {
-                ecs.Undo(currentThread);
+                // Re-enregistrer variables post EH with 1 post-fix so they can be used in registers rather than from stack
+                SynchronizationContext previousSyncCtx1 = previousSyncCtx;
+                Thread currentThread1 = currentThread;
+                SynchronizationContext currentSyncCtxRef1 = currentThread1.SynchronizationContext;
+                // The common case is that these have not changed, so avoid the cost of a write barrier if not needed.
+                if (previousSyncCtx1 != currentSyncCtxRef1)
+                {
+                    // Restore changed SynchronizationContext back to previous
+                    currentSyncCtxRef1 = previousSyncCtx1;
+                }
+
+                ExecutionContext previousExecutionCtx1 = previousExecutionCtx;
+                ExecutionContext currentExecutionCtxRef1 = currentThread1.ExecutionContext;
```

Comment:
It looks like you are missing a `ref`. But then it's not clear what this is needed for to begin with.

Reply:
Didn't push the changes :(

Reply:
Hmm, I'm not sure why you're using `ref` here. It may not have the outcome you think it does.

Reply:
For the restore a few lines down?
```csharp
// Restore changed ExecutionContext back to previous
currentExecutionCtxRef1 = previousExecutionCtx1;
```

Reply:
Why not just `currentThread1.ExecutionContext = previousExecutionCtx1;`?

Reply:
Changing to
```csharp
ExecutionContext currentExecutionCtx1 = currentThread1.ExecutionContext;
if (previousExecutionCtx1 != currentExecutionCtx1)
{
    // Restore changed ExecutionContext back to previous
    currentThread1.ExecutionContext = previousExecutionCtx1;
```
Changes asm from
```asm
G_M42258_IG11:
       mov      rdi, gword ptr [rbp-20H]
       lea      rcx, bword ptr [rsi+8]
       mov      rsi, gword ptr [rcx]
       cmp      rdi, rsi
       je       SHORT G_M42258_IG14
       mov      rdx, rdi
       call     CORINFO_HELP_CHECKED_ASSIGN_REF
       test     rsi, rsi
       je       SHORT G_M42258_IG12
       cmp      gword ptr [rsi+16], 0
       jne      SHORT G_M42258_IG13
```
To
```asm
G_M42259_IG11:
       mov      rdi, gword ptr [rbp-28H]
       mov      rbx, gword ptr [rsi+8]
       cmp      rdi, rbx
       je       SHORT G_M42259_IG14
       lea      rcx, bword ptr [rsi+8]
       mov      rdx, rdi
       call     CORINFO_HELP_ASSIGN_REF
       test     rbx, rbx
       je       SHORT G_M42259_IG12
       cmp      gword ptr [rbx+16], 0
       jne      SHORT G_M42259_IG13
```
Does look better(?)

Reply:
Same for Run then?

Reply:
> Does look better(?)

Yes, a `lea` instruction is now executed only when needed and you get a normal GC write barrier instead of a checked one. In general, I can't think of any case where taking a ref of a class field could be useful when the class object reference is already a local.

Reply:
> Same for Run then?

If it has the same effected on the generated code, sure.

Reply:
> In general, I can't think of any case where taking a ref of a class field could be useful when the class object reference is already a local. 

I think it was a side effect of trying to improve the code when everything was stack based

Reply:
Yes, likely.